### PR TITLE
MINOR: Extend mirror maker test to include interceptors

### PR DESCRIPTION
**Pull request by kkonstantine**

### CODE REVIEW

**Comment by kkonstantine** [comment]
This PR allows MM to accept custom interceptors. @ewencp @hachikuji I'd appreciate your reviews.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala [hunk 1]
```diff
@@ -701,7 +701,7 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
   private[tools] object defaultMirrorMakerMessageHandler extends MirrorMakerMessageHandler {
     override def handle(record: BaseConsumerRecord): util.List[ProducerRecord[Array[Byte], Array[Byte]]] = {
       val timestamp: java.lang.Long = if (record.timestamp == Record.NO_TIMESTAMP) null else record.timestamp
-      Collections.singletonList(new ProducerRecord[Array[Byte], Array[Byte]](record.topic, null, timestamp, record.key, record.value))
+      Collections.singletonList(new ProducerRecord[Array[Byte], Array[Byte]](record.topic, record.partition, timestamp, record.key, record.value))
```
Comment (ewencp) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: Why are we changing MM's behavior in this patch? This certainly wouldn't be covered under what the commit message says.

More importantly, this would almost certainly break things for some users as they may be using a different number of partitions in source and destination clusters. The default handler preserves ordering based on keys but allows for a different number of partitions. Normally a different `MirrorMakerMessageHandler` would be swapped in if the user wanted different behavior.

File: tests/kafkatest/services/connect.py [hunk 2]
```diff
@@ -57,6 +57,7 @@ def __init__(self, context, num_nodes, kafka, files):
         self.kafka = kafka
         self.security_config = kafka.security_config.client_config()
         self.files = files
+        self.kafka_heap_opts = None
```
Comment (ewencp) [inline tests/kafkatest/services/connect.py]: Should this just be more generic? Seems like we could end up adding flags for a ton of different envvars. Is there any issue with having the caller set the envvars more directly since they have to know what the available settings are anyway?

File: tests/kafkatest/services/templates/mirror_maker_consumer.properties [hunk 3]
```diff
@@ -27,3 +29,9 @@ offsets.storage={{ offsets_storage }}
 {% if consumer_timeout_ms is defined and consumer_timeout_ms is not none %}
 consumer.timeout.ms={{ consumer_timeout_ms }}
 {% endif %}
+
+{% if new_consumer %}
```
Comment (ewencp) [inline tests/kafkatest/services/templates/mirror_maker_consumer.properties]: This change seems unrelated to the patch? And changes the defaults such that no tests would cover the default MM settings?

File: tests/kafkatest/services/mirror_maker.py [hunk 4]
```diff
@@ -108,11 +108,18 @@ def __init__(self, context, num_nodes, source, target, whitelist=None, blacklist
             raise Exception("offsets_storage should be 'kafka' or 'zookeeper'. Instead found %s" % self.offsets_storage)
 
         self.offset_commit_interval_ms = offset_commit_interval_ms
+        self.interceptor_classes = interceptor_classes
```
Comment (ewencp) [inline tests/kafkatest/services/mirror_maker.py]: Are these for producer or consumer? In MM you could potentially have both. It would be good to avoid ambiguity here, even if only one of the two is currently needed.

File: tests/kafkatest/services/mirror_maker.py [hunk 5]
```diff
@@ -108,11 +108,18 @@ def __init__(self, context, num_nodes, source, target, whitelist=None, blacklist
...
+        self.external_jars = None
 
     def start_cmd(self, node):
         cmd = "export LOG_DIR=%s;" % MirrorMaker.LOG_DIR
         cmd += " export KAFKA_LOG4J_OPTS=\"-Dlog4j.configuration=file:%s\";" % MirrorMaker.LOG4J_CONFIG
         cmd += " export KAFKA_OPTS=%s;" % self.security_config.kafka_opts
+        cmd += " export KAFKA_OPTS=%s;" % self.security_config.kafka_opts
+        # add external dependencies, for instance for interceptors
+        if self.external_jars is not None:
+            cmd += "for file in %s; do CLASSPATH=$CLASSPATH:$file; done; " % self.external_jars
```
Comment (ewencp) [inline tests/kafkatest/services/mirror_maker.py]: Does this work properly for both lists of individual files and globs?

File: tests/kafkatest/services/templates/mirror_maker_consumer.properties [hunk 6]
```diff
@@ -16,9 +16,11 @@
 
 {% if new_consumer %}
 bootstrap.servers={{ source.bootstrap_servers(security_config.security_protocol) }}
+auto.offset.reset={{ source_auto_offset_reset|default('latest') }}
```
Comment (ewencp) [inline tests/kafkatest/services/templates/mirror_maker_consumer.properties]: Also seems unrelated? And this setting doesn't appear in the service anywhere? I think it's fine to have sections of the template that use extra fields if they are assigned, but then they probably need docs somewhere. It'd probably be good to standardize on either passing all parameters through the constructor (even if it uses `**kwargs` to avoid enumerating them all in the constructor's signature) or assigning fields on the Service object after it is created.

File: core/src/main/scala/kafka/tools/MirrorMaker.scala (hunk 1 above)
Comment (kkonstantine) [reply core/src/main/scala/kafka/tools/MirrorMaker.scala]: I agree, this should be a breaking change. I'm reverting.

File: tests/kafkatest/services/connect.py (hunk 2 above)
Comment (kkonstantine) [reply tests/kafkatest/services/connect.py]: Fair point. I added variables based on existing KAFKA_\* env vars. This more generic way will add some burden to the user w.r.t expressing the entire set of variables as a string, but it's not unreasonable.

File: tests/kafkatest/services/mirror_maker.py (hunk 4 above)
Comment (kkonstantine) [reply tests/kafkatest/services/mirror_maker.py]: It's the producer's. I just noticed that `interceptor.classes` property was recently split between producer and consumer classes with a prefix. I'm inclined to add only what's needed at this point, which is producer classes.

File: tests/kafkatest/services/mirror_maker.py (hunk 5 above)
Comment (kkonstantine) [reply tests/kafkatest/services/mirror_maker.py]: Globs meaning wildcards? I believe so. It's used in the same way as in ProducerPerformanceService for instance.

File: tests/kafkatest/services/templates/mirror_maker_consumer.properties (hunk 3 above)
Comment (kkonstantine) [reply tests/kafkatest/services/templates/mirror_maker_consumer.properties]: I was getting a warning without this setting, and I was thinking that the default which was having only range based assignor is as bad. But I agree. I'll see if my interceptors work without this change ported here.

Comment (kkonstantine) [reply tests/kafkatest/services/templates/mirror_maker_consumer.properties]: This corresponds to the defaults now.

File: tests/kafkatest/services/templates/mirror_maker_consumer.properties (hunk 6 above)
Comment (kkonstantine) [reply tests/kafkatest/services/templates/mirror_maker_consumer.properties]: The field was added in the constructor body to facilitate self-documentation of this class

File: tests/kafkatest/services/mirror_maker.py [hunk 7]
```diff
@@ -108,11 +108,22 @@ def __init__(self, context, num_nodes, source, target, whitelist=None, blacklist
...
+
+        # These properties are potentially used by third-party tests.
+        self.source_auto_offset_reset = None
+        self.partition_assignment_strategy = None
 
     def start_cmd(self, node):
         cmd = "export LOG_DIR=%s;" % MirrorMaker.LOG_DIR
         cmd += " export KAFKA_LOG4J_OPTS=\"-Dlog4j.configuration=file:%s\";" % MirrorMaker.LOG4J_CONFIG
         cmd += " export KAFKA_OPTS=%s;" % self.security_config.kafka_opts
+        cmd += " export KAFKA_OPTS=%s;" % self.security_config.kafka_opts
```
Comment (ewencp) [inline tests/kafkatest/services/mirror_maker.py]: This line is duplicated

File: tests/kafkatest/services/connect.py [hunk 8]
```diff
@@ -208,6 +208,8 @@ def node(self):
     def start_cmd(self, node, connector_configs):
         cmd = "( export KAFKA_LOG4J_OPTS=\"-Dlog4j.configuration=file:%s\"; " % self.LOG4J_CONFIG_FILE
         cmd += "export KAFKA_OPTS=%s; " % self.security_config.kafka_opts
+        for envvar in self.environment:
```
Comment (ewencp) [inline tests/kafkatest/services/connect.py:212]: `self.environment` may not be defined, e.g. it isn't in any of the connect tests.

Comment (kkonstantine) [reply tests/kafkatest/services/connect.py:212]: Update: I initialize it in the base constructor as an empty dictionary

File: tests/kafkatest/services/templates/mirror_maker_consumer.properties [hunk 9]
```diff
@@ -16,9 +16,11 @@
...
 {% else %}
 zookeeper.connect={{ source.zk.connect_setting() }}
 zookeeper.connection.timeout.ms={{ zookeeper_connection_timeout_ms|default(6000) }}
+auto.offset.reset={{ source_auto_offset_reset|default('largest') }}
```
Comment (ewencp) [inline tests/kafkatest/services/templates/mirror_maker_consumer.properties]: Uggh, I forgot this about jinja2 -- `default` only applies if the value is not defined, whereas `None` counts as a value. That is why elsewhere we have `if` statements like `if x is defined and x is not none`. So these are actually getting `None` instead of the default with the current code.

Not sure which we want to prefer here. The options are either take advantage of `default` and get concision in the template but leave things undocumented in the class (i.e. to use the flag, you need to know to go look at the template files); or use the the more verbose if statements and wrap them. There are also some other possibilities suggested [here](http://stackoverflow.com/questions/19614027/jinja2-template-variable-if-none-object-set-a-default-value), only doing the inline `if`/`else` with explicit check for `None` seems to work in all cases afaict...

**Comment by ewencp** [comment]
Changes LGTM and we seem to have gotten a clean run of the tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/140/

**Comment by kkonstantine** [comment]
Thanks for noticing the successful run (I just wanted to mention today) and for the review!
